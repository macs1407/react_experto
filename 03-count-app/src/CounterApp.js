import React, { useState } from 'react';
import PropTypes from 'prop-types';


const CounterApp = ({numero = 5}) => {

    // Uso de hook useState, un hook no es mas que una funcion
    // en este caso useState tiene un nombre y una funcion que apunta a ese nombre
    // El useState retorna un arreglo con 2 valores
    const [counter, setCounter] = useState(numero)

    const handleApp = (e)=>{
        // No se puede poner counter++ por que sria lo mismo que decir
        // counter = counter + 1 y se estaria intentando cambiar el state de una constante
        setCounter(counter+1);
    } 

    const handleReset = (e) =>{
        setCounter(numero);
    }

    const handlerMinus = (e) =>{
        setCounter(counter-1);
    }

    return ( 
        <>
            <h1>Counter app</h1>
            <h2>{numero}</h2>
            <h3>{counter}</h3>
            {/* No se pone handleApp() se pone handleApp, por entonces ejecutaria la funcion cuando se 
            renderice el componente y ese no es el comportamiento adecuado */}
            <button onClick={ handleApp }>+1</button>
            <button onClick={ handleReset }>Reset</button>
            <button onClick={ handlerMinus }>-1</button>
        </>
     );
}

CounterApp.propTypes = {
    numero: PropTypes.number
}
 
export default CounterApp;