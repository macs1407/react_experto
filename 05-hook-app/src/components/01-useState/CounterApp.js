import React, { useState } from 'react';
import './counter.css';

export const CounterApp = () => {

    const [counter, setCounter] = useState({
        counterUno:5,
        counterDos:30,
        counterTres:10,
    });

    const {counterUno, counterDos} = counter;

    const handleChange = ()=>{
        // Para dejar una copia del state actual se hace con el operador expres ...
        // y se modifica solo la parte que nos interesa en este caso counterUno
        setCounter({
            ...counter,
            counterUno: counterUno+1
        })
    }

    return (
        <div>
            <h1>Counter Uno {counterUno}</h1>
            <h1>Counter Dos {counterDos}</h1>
            <hr />
            <button className="btn btn-primary" onClick={ handleChange }>+1</button>
        </div>
    )
}
