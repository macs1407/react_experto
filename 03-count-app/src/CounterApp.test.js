import React from 'react';
import '@testing-library/jest-dom';
import {shallow} from 'enzyme';
import CounterApp from './CounterApp';

describe("Pruebas del componente <CounterApp />", ()=>{
    let wraper = shallow(<CounterApp />);

    // Se ejecuta antes de cada test
    beforeEach(()=>{
        wraper = shallow(<CounterApp />);
    });

    test('Comprobar que el componente se renderice de forma correcta ', () => {
        const numero = 30;
        const wraper = shallow(<CounterApp numero={numero} />);
        expect(wraper).toMatchSnapshot();
    });
    
    test('Debe de mostrar el valor por defecto 100 ', () => {
        
        const texto = wraper.find("h3").text();
        console.log(texto);
        expect(texto).toBe("5");
    });

    test('Debe de incrementar el valor mas uno ', () => {
        // Buscar el primer boton de incrementar
        wraper.find('button').at(0).simulate("click");
        const texto = wraper.find("h3").text();
        expect(texto).toBe("6");
    });

    test('Debe de dejar el valor como esta inicialmente ', () => {
        wraper.find('button').at(0).simulate("click");
        wraper.find('button').at(0).simulate("click");
        let texto = wraper.find("h3").text();
        console.log(texto);
        // Buscar el primer boton de dejar por defecto
        wraper.find('button').at(1).simulate("click");
        texto = wraper.find("h3").text();
        expect(texto).toBe("5");
    });

    test('Debe de decrementar el valor de uno ', () => {
        // Buscar el primer boton de decrementar
        wraper.find('button').at(2).simulate("click");
        const texto = wraper.find("h3").text();
        expect(texto).toBe("4");
    });
});