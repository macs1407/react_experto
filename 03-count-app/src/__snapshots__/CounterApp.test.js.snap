// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <CounterApp /> Comprobar que el componente se renderice de forma correcta  1`] = `
<Fragment>
  <h1>
    Counter app
  </h1>
  <h2>
    30
  </h2>
  <h3>
    30
  </h3>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    -1
  </button>
</Fragment>
`;
